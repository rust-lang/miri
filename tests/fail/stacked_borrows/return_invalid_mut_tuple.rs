// Make sure that we cannot return a `&mut` that got already invalidated, not even in a tuple.
// Due to shallow reborrowing, the error only surfaces when we look into the tuple.
fn foo(x: &mut (i32, i32)) -> (&mut i32,) {
    let xraw = x as *mut (i32, i32);
    let ret = (unsafe { &mut (*xraw).1 },);
    let _val = unsafe { *xraw }; // invalidate xref
    ret
}

fn main() {
    foo(&mut (1, 2)).0; //~ ERROR: /retag .* tag does not exist in the borrow stack/
}
