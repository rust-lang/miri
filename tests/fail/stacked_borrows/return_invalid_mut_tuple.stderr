error: Undefined Behavior: trying to retag from <TAG> for Unique permission at ALLOC[0x4], but that tag does not exist in the borrow stack for this location
  --> $DIR/return_invalid_mut_tuple.rs:LL:CC
   |
LL |     foo(&mut (1, 2)).0;
   |     ^^^^^^^^^^^^^^^^^^
   |     |
   |     trying to retag from <TAG> for Unique permission at ALLOC[0x4], but that tag does not exist in the borrow stack for this location
   |     this error occurs as part of retag at ALLOC[0x4..0x8]
   |
   = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <TAG> was created by a Unique retag at offsets [0x4..0x8]
  --> $DIR/return_invalid_mut_tuple.rs:LL:CC
   |
LL |     let ret = (unsafe { &mut (*xraw).1 },);
   |                         ^^^^^^^^^^^^^^
help: <TAG> was later invalidated at offsets [0x0..0x8] by a read access
  --> $DIR/return_invalid_mut_tuple.rs:LL:CC
   |
LL |     let _val = unsafe { *xraw }; // invalidate xref
   |                         ^^^^^
   = note: BACKTRACE:
   = note: inside `main` at $DIR/return_invalid_mut_tuple.rs:LL:CC

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error

